jmix-rest/rest/src/main/java/io/jmix/rest/impl/service/EntitiesControllerManager.java

```diff
@@ -193,35 +201,33 @@ public EntitiesSearchResult searchEntities(String entityName,
         MetaClass metaClass = restControllerUtils.getMetaClass(entityName);
         checkCanReadEntity(metaClass);
 
-        RestFilterParseResult filterParseResult;
+        Condition jmixCondition;
         try {
-            filterParseResult = restFilterParser.parse(filterJson, metaClass);
+            jmixCondition = restFilterParser.parse(filterJson, metaClass);
         } catch (RestFilterParseException e) {
             throw new RestAPIException("Cannot parse entities filter", e.getMessage(), HttpStatus.BAD_REQUEST, e);
         }
 
-        String jpqlWhere = filterParseResult.getJpqlWhere();
-        Map<String, Object> queryParameters = filterParseResult.getQueryParameters();
-
-        String queryString = "select e from " + entityName + " e";
+        String json = _loadEntitiesList(jmixCondition, viewName, limit, offset, sort, returnNulls,
+                dynamicAttributes, modelVersion, metaClass);
+        Long count = BooleanUtils.isTrue(returnCount) ?
+                countEntities(metaClass, jmixCondition)
+                : null;
+        return new EntitiesSearchResult(json, count);
+    }
 
-        if (jpqlWhere != null) {
-            queryString += " where " + jpqlWhere.replace("{E}", "e");
-        }
+    protected long countEntities(MetaClass metaClass,@Nullable Condition jmixCondition) {
+        String queryString = createSimpleSelect(metaClass);
+        queryString = QueryUtils.applyQueryStringProcessors(queryStringProcessors, queryString, metaClass.getJavaClass());
 
-        String json = _loadEntitiesList(queryString, viewName, limit, offset, sort, returnNulls,
-                dynamicAttributes, modelVersion, metaClass, queryParameters);
-        Long count = null;
-        if (BooleanUtils.isTrue(returnCount)) {
-            LoadContext ctx = new LoadContext(metadata.getClass(metaClass.getJavaClass()))
-                    .setQuery(new LoadContext.Query(queryString));
-            if (queryParameters != null) {
-                ctx.getQuery().setParameters(queryParameters);
-            }
-            count = dataManager.getCount(ctx);
-        }
+        LoadContext<?> ctx = new LoadContext<>(metadata.getClass(metaClass.getJavaClass()))
+                .setQuery(new LoadContext.Query(queryString)
+                        .setCondition(jmixCondition));
+        return dataManager.getCount(ctx);
+    }
 
-        return new EntitiesSearchResult(json, count);
+    protected static String createSimpleSelect(MetaClass metaClass){
+        return "select e from " + metaClass.getName() + " e";
```

Comment:
Return JPQL query only for JPA entities (`MetadataTools.isJpaEntity()`). Otherwise return empty string. See for example `io.jmix.restds.impl.RestDataStore.countAll` - it ignores the query text.

Reply:
Fixed, but null is used instead of empty string because it is more clear way to say that query is absent and `io.jmix.core.LoadContext.Query#getQueryString()` allow nulls now.